server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,383 @@
+                @Override
+                public void onResponse(DiscoveryNode remoteNode) {
+                    assert remoteNode.isMasterNode() : remoteNode + " is not master-eligible";
+                    assert remoteNode.equals(getLocalNode()) == false : remoteNode + " is the local node";
+                    synchronized (mutex) {
+                        if (active) {
+                            assert discoveryNode.get() == null : "discoveryNode unexpectedly already set to " + discoveryNode.get();
+                            discoveryNode.set(remoteNode);
+                            requestPeers();
+                        }
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.debug(() -> new ParameterizedMessage("{} connection failed", Peer.this), e);
+                    removePeer();
+                }
+            });
+        }
+
+        private void removePeer() {
+            final Peer removed = peersByAddress.remove(transportAddress);
+            assert removed == Peer.this;
+        }
+
+        private void requestPeers() {
+            assert holdsLock() : "PeerFinder mutex not held";
+            assert peersRequestInFlight == false : "PeersRequest already in flight";
+            assert active;
+
+            final DiscoveryNode discoveryNode = getDiscoveryNode();
+            assert discoveryNode != null : "cannot request peers without first connecting";
+
+            logger.trace("{} requesting peers from {}", this, discoveryNode);
+            peersRequestInFlight = true;
+
+            List<DiscoveryNode> knownNodes = getKnownPeers();
+
+            transportService.sendRequest(discoveryNode, REQUEST_PEERS_ACTION_NAME,
+                new PeersRequest(getLocalNode(), knownNodes),
+                new TransportResponseHandler<PeersResponse>() {
+
+                    @Override
+                    public PeersResponse read(StreamInput in) throws IOException {
+                        return new PeersResponse(in);
+                    }
+
+                    @Override
+                    public void handleResponse(PeersResponse response) {
+                        logger.trace("{} received {} from {}", Peer.this, response, discoveryNode);
+                        synchronized (mutex) {
+                            if (active == false) {
+                                return;
+                            }
+
+                            peersRequestInFlight = false;
+
+                            if (response.getMasterNode().isPresent()) {
+                                final DiscoveryNode masterNode = response.getMasterNode().get();
+                                if (masterNode.equals(discoveryNode) == false) {
+                                    startProbe(masterNode.getAddress());
+                                }
+                            } else {
```

Comment:
I wonder if we should even make this distinction. I think the following should suffice:

```
response.getMasterNode().map(DiscoveryNode::getAddress).ifPresent(PeerFinder.this::startProbe);
response.getKnownPeers().stream().map(DiscoveryNode::getAddress).forEach(PeerFinder.this::startProbe);
```

Reply:
It's all about the one-liners today. See 221253a3f656a1e47d28ff30c551e33032be5a61.